iam/role/role.go

```diff
@@ -0,0 +1,115 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/acorn-io/services/aws/libs/common"
+	"github.com/aws/aws-cdk-go/awscdk/v2"
+	"github.com/aws/aws-cdk-go/awscdk/v2/awsiam"
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/constructs-go/constructs/v10"
+	"github.com/aws/jsii-runtime-go"
+	"github.com/sirupsen/logrus"
+)
+
+type IAMRoleStackProps struct {
+	StackProps                awscdk.StackProps
+	Tags                      map[string]string      `json:"tags"`
+	RoleName                  string                 `json:"roleName"`
+	TrustedArn                string                 `json:"trustedArn"`
+	Policy                    map[string]interface{} `json:"policy"`
+	MaxSessionDurationMinutes int                    `json:"maxSessionDurationMinutes"`
+	Path                      string                 `json:"path"`
+	ExternalIds               string                 `json:"externalIds"`
+	Description               string                 `json:"description"`
+}
+
+func (rsp *IAMRoleStackProps) setDefaults() {
+	if rsp.RoleName == "" {
+		rsp.RoleName = os.Getenv("ACORN_EXTERNAL_ID")
+	}
+	if rsp.MaxSessionDurationMinutes == 0 {
+		rsp.MaxSessionDurationMinutes = 60
+	}
+	if rsp.Path == "" {
+		rsp.Path = "/"
+	}
+}
+
+func (rsp *IAMRoleStackProps) validateProps() error {
+	var errs []error
+	if rsp.MaxSessionDurationMinutes < 60 {
+		errs = append(errs, fmt.Errorf("maxSessionDurationMinutes must be at least 60"))
+	}
+	if !strings.HasPrefix(rsp.Path, "/") {
+		errs = append(errs, fmt.Errorf("path must start with a /"))
+	}
+	if len(rsp.Policy) == 0 {
+		errs = append(errs, fmt.Errorf("policy cannot be empty"))
+	}
+	if _, err := arn.Parse(rsp.TrustedArn); err != nil {
+		errs = append(errs, fmt.Errorf("failed to parse trustedArn: %w", err))
+	}
+	return errors.Join(errs...)
+}
+
+func newIAMRoleStack(scope constructs.Construct, id string, props *IAMRoleStackProps) (awscdk.Stack, error) {
+	var sprops awscdk.StackProps
+	if props != nil {
+		sprops = props.StackProps
+	}
+	stack := awscdk.NewStack(scope, &id, &sprops)
+
+	roleProps := &awsiam.RoleProps{
+		AssumedBy:          awsiam.NewArnPrincipal(jsii.String(props.TrustedArn)),
```

Comment:
I believe this field is for a service principle like `ec2.amazonaws.com`, not an ARN.

Reply:
Looking at some docs, it seems like its supposed to be a service principle + a trust policy document.

Reply:
So the `AssumedBy` field can be a number of different things, but it has to be some kind of principal. In this case, the `ArnPrincipal` is pretty flexible and can take ARNs for a number of different things: https://pkg.go.dev/github.com/aws/aws-cdk-go/awscdk/v2/awsiam@v2.93.0#ArnPrincipal

I already know from my testing that specifying the ARN of another Role here works.

Reply:
Interesting, so they have some helpers around this. Cool beans